core/src/main/resources/lib/form/class-entry.jelly

```diff
@@ -36,12 +46,14 @@ THE SOFTWARE.
   </st:documentation>
   <j:set var="clazz" value="${attrs.clazz ?: attrs.descriptor.clazz.name}" />
   <f:invisibleEntry>
-    <!-- Legacy: Remove once plugins have been staged onto $class -->
-    <input type="hidden" name="stapler-class" value="${clazz}" />
-    <!-- Legacy: Remove once plugins have been staged onto $class -->
-    <j:if test="${attrs.descriptor != null}">
-      <input type="hidden" name="kind" value="${attrs.descriptor.id}" />
-    </j:if>
-    <input type="hidden" name="$class" value="${clazz}" />
+    <j:choose>
+      <j:when test="${descriptor != null and descriptor.id != clazz}">
+        <input type="hidden" name="kind" value="${attrs.descriptor.id}" />
+      </j:when>
```

Comment:
Same for [JENKINS-28110 Performance plugin problem](https://issues.jenkins-ci.org/browse/JENKINS-28110).
performance-plugin ovierrides `Descriptor#getId` for almost all its descriptors (overriding in the base class https://github.com/jenkinsci/performance-plugin/blob/performance-1.12/src/main/java/hudson/plugins/performance/PerformanceReportParserDescriptor.java#L16).